convert2rhel/backup/subscription.py

```diff
@@ -0,0 +1,137 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright(C) 2016 Red Hat, Inc.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+__metaclass__ = type
+
+import logging
+import re
+
+from convert2rhel import subscription, utils
+from convert2rhel.backup import RestorableChange
+
+
+loggerinst = logging.getLogger(__name__)
+
+
+class RestorableSystemSubscription(RestorableChange):
+    """
+    Register with RHSM in a fashion that can be reverted.
+    """
+
+    # We need this __init__ because it is an abstractmethod in the base class
+    def __init__(self):  # pylint: disable=useless-parent-delegation
+        super(RestorableSystemSubscription, self).__init__()
+
+    def enable(self):
+        """Register and attach a specific subscription to OS."""
+        if self.enabled:
+            return
+
+        subscription.register_system()
+        subscription.attach_subscription()
+
+        super(RestorableSystemSubscription, self).enable()
+
+    def restore(self):
+        """Rollback subscription related changes"""
+        loggerinst.task("Rollback: RHSM-related actions")
+
+        if self.enabled:
+            try:
+                subscription.unregister_system()
+            except subscription.UnregisterError as e:
+                loggerinst.warning(str(e))
+            except OSError:
+                loggerinst.warning("subscription-manager not installed, skipping")
+
+        super(RestorableSystemSubscription, self).restore()
+
+
+class RestorableAutoAttachmentSubscription(RestorableChange):
+    """
+    Auto attach subscriptions with RHSM in a fashion that can be reverted.
+    """
+
+    def __init__(self):
+        super(RestorableAutoAttachmentSubscription, self).__init__()
+        self._is_attached = False
+
+    def enable(self):
+        self._is_attached = subscription.auto_attach_subscription()
+        super(RestorableAutoAttachmentSubscription, self).enable()
+
+    def restore(self):
+        if self._is_attached:
+            subscription.remove_subscription()
```

Comment:
```suggestion
            loggerinst.task("Rollback: De-auto-attaching subscription")
            subscription.remove_subscription()
```
Or some similar message